Fix issue #23774 Make transport_test use exec

smikulcik opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Since ssh may re-route traffic to any number of places before it 
reaches a host by the ssh config files and proxy hosts, we cannot use
sockets to test our transport.

It would make sense to use the same mechanism that we use to
perform actions on a host as we would to verify connectivity.

This change performs a dummy exec_command on the host to verify
that its transport works instead of using a single socket.

In particular, this fixes https://github.com/ansible/ansible/issues/23774

By using the ssh configuration built into exec_command, we are able to
handle proxy jumps specified in ssh config files.  This allows us to handle
bastion hosts when using wait_for_connection

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
wait_for_connection
ansible.plugins.connection.accelerate
ansible.plugins.connection.ssh
ansible.plugins.connection.paramiko_ssh
ansible.plugins.connection.winrm

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0
  config file = 
  configured module search path = [u'/Users/smikulc/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/smikulc/virtualenvs/ansible/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible
  executable location = /Users/smikulc/virtualenvs/ansible/bin/ansible
  python version = 2.7.10 (default, Jul 30 2016, 19:40:32) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)]
```


##### ADDITIONAL INFORMATION

For testing purposes, here is a demo that illustrates the bug fixed in this change https://github.com/smikulcik/ansible-waitforconnection-bug-demo

dagwieers commented on lib/ansible/plugins/connection/accelerate.py:
```diff
@@ -100,11 +100,9 @@ def _connect(self):
-        display.vvv("attempting transport test to %s:%s" % (host, port))
-        sock = socket.create_connection((host, port), connect_timeout)
-        sock.close()
+        display.vvv("attempting transport test to %s" % self.host)
```
Fails because self_host is not defined for accelerate transport.

dagwieers commented on lib/ansible/plugins/connection/accelerate.py:
```diff
@@ -100,11 +100,9 @@ def _connect(self):
-        sock.close()
+        display.vvv("attempting transport test to %s" % self.host)
+        self.reset()
+        self.exec_command("exit")
```
I am not sure if exit is an accepted command.

dagwieers commented on lib/ansible/plugins/connection/paramiko_ssh.py:
```diff
@@ -134,11 +134,9 @@ class Connection(ConnectionBase):
-        display.vvv("attempting transport test to %s:%s" % (host, port))
-        sock = socket.create_connection((host, port), connect_timeout)
-        sock.close()
+        display.vvv("attempting transport test to %s" % self.host)
```
Here also self.host is not defined.

dagwieers commented on lib/ansible/plugins/connection/winrm.py:
```diff
@@ -89,11 +89,9 @@ def __init__(self,  *args, **kwargs):
-        display.vvv("attempting transport test to %s:%s" % (host, port))
-        sock = socket.create_connection((host, port), connect_timeout)
-        sock.close()
+        display.vvv("attempting transport test to %s" % self.host)
```
Again self_host is not defined for winrm.

dagwieers commented on lib/ansible/plugins/connection/winrm.py:
```diff
@@ -89,11 +89,9 @@ def __init__(self,  *args, **kwargs):
-        sock.close()
+        display.vvv("attempting transport test to %s" % self.host)
+        self.reset()
+        self.exec_command("exit")
```
I would probably go for "Exit" here, but that's more cosmetic than functional.

dagwieers reviewed (changes requested):
We wanted to avoid doing a full roundtrip using the protocol as it is more expensive than a single TCP test. So maybe we should modify the default delay for these tests.

We could also get rid of the transport_test altogether, and only do the real remote module run.

smikulcik committed:
Fix issue #23774 Make transport_test use exec

Some ansible hosts are not directly reachable
via socket connections.  It is best to use
the transport mechanism itself than to assume
that the host is reachable via socket.

If we can exec a command, then the transport
is working.

smikulcik commented:
@dagwieers Thanks for the feedback. What would you recommend for configuring connect_timeout?  Currently, the exec_command uses the same timeout from the self._play_context.timeout, this code is ignoring anything set for the connect_timeout.

dagwieers commented:
@smikulcik Good question, I don't know what we do best in this case. Maybe we should discuss this in the core meeting.

dagwieers commented:
The integration test found a weird issue related to our code:
```
timed out waiting for connection port up: list index out of range
```
Needs investigation.

dagwieers commented:
Beware that the deadline for getting new features/modules accepted in [Ansible v2.5 is nearing](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_5.rst), it is set to either 2018-01-15 or 2018-01-31. If you are blocked, or you need feedback, please discuss on IRC channel #ansible-windows or add a comment to [the Windows Working Group meeting agenda]

dagwieers commented:
I still think this would be a useful addition, the transport test was disabled so it would still work over proxied connections, but I prefer keeping this around.

ShachafGoldstein commented:
@dagwieers @jborean93 @nitzmahone - This one seems dead for quite a while, still relevant?

jborean93 commented:
Considering the age and merge conflicts I am going to close this PR. If this is still a problem please feel free to open a new PR that has been rebased.
